File: LeetCode/Medium/0036. Valid Sudoku.java
class Solution {
    public boolean checkValid(char[] check) {
        for(int i = 0; i < 9; ++i) {
            for(int j = i+1; j < 9; ++j) {
                if(check[i] == check[j] && check[i] != '.') {
                    return false;
                }
            }
        }
        return true;
    }


    public boolean isValidSudoku(char[][] board) {
        char[] check = new char[9];

        // check all rows
        for(int row = 0; row < 9; ++row) {
            for(int i = 0; i < 9; ++i) {
                check[i] = board[row][i];
            }
            if(!checkValid(check)) return false;
        }

        // check all cols
        for(int col = 0; col < 9; ++col) {
            for(int j = 0; j < 9; ++j) {
                check[j] = board[j][col];
            }
            if(!checkValid(check)) return false;
        }

        // check all subsquares
        for(int startRow = 0; startRow < 9; startRow+=3) {
            for(int startCol = 0; startCol < 9; startCol+=3) {
                // check individual subsquare
                int ind = 0;
                for(int i = startRow; i < startRow+3; ++i) {
                    for(int j = startCol; j < startCol+3; ++j) {
                        check[ind] = board[i][j];
                        ++ind;
                    }
                }
                if(!checkValid(check)) return false;
            }
        }

        return true;
    }
}

File: LeetCode/Medium/0036. Valid Sudoku.py
BOARD_LEN = 9
SQR_LEN = 3

class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool

        algo:
        for each coordinate: check if it's valid in its row, col & square
        """
        cols = collections.defaultdict(set)
        rows = collections.defaultdict(set)
        squares = collections.defaultdict(set)

        for i in range(0, BOARD_LEN):
            for j in range(0, BOARD_LEN):
                if board[i][j] != ".":
                    if (board[i][j] in rows[i]) or (board[i][j] in cols[j]) or(board[i][j] in squares[(i // SQR_LEN, j // SQR_LEN)]):
                        return False
                    else:
                        rows[i].add(board[i][j])
                        cols[j].add(board[i][j])
                        squares[(i // SQR_LEN, j // SQR_LEN)].add(board[i][j])
    
        return True

    
